### ISSUE TITLE
FR: PDF/DJVU line text selection behavior

### ISSUE THREAD

**Issue by jonnyl2** [koreader/koreader#12753]
**Does your feature request involve difficulty completing a task? Please describe.**

On reflowable documents, selecting text up to the last word of a paragraph is very convenient, as even if the last word ends at the beginning of the line, KOReader will automatically select all words on the line panned to (up to the `x`-position):

https://github.com/user-attachments/assets/c02896fb-3358-44a0-9061-6f4b68b8438e

On PDF/DJVU files on the other hand, nothing on the same line is being selected when panning to the empty part of a line (only once parts of the following line are included). To select the end of the sentence, it is necessary to pan over to the last word of the line:

https://github.com/user-attachments/assets/67058c17-ea8e-45c3-91bd-c45eb1f4585d

**Describe the solution you'd like**

Harmonize the text selection behavior in PDFs such that it works similarly as in EPUBs: select all previous words on the line panned to, even when panning to a blank part of the line. 

**Describe alternatives you've considered**

\--

**Additional context**

\--

**Comment by hius07** [comment]
In epud it's easy, we have "lines".
We can highlight `impossible` when long-pressing the red dot, or highlight the full line when panning to it.

![1](https://github.com/user-attachments/assets/954f52fa-12b9-469f-b514-82a332c9abf6)

In pdf there are no "lines", we have "boxes". They can be placed anywhere in the page.

![2](https://github.com/user-attachments/assets/2bf3a7a4-1997-4316-a8b7-2cd694b2e877)

When long-pressing the small dot, we can select `replacement`. What should we highlight when long-pressing the big dot?

Should we select the line when panning to the red dots?
For the small dot - yes, for the big dot - no. We cannot distinguish.

So we go the safe way: look for the **nearest** box (`main` for the bigger dot) and highlight it and all the previous boxes.

**Comment by jonnyl2** [comment]
>When long-pressing the small dot, we can highlight replacement. 

Yes, that would work. There wouldn't really be any reason for a user to long-press or pan to the big dot. (IMO, nothing at all would have to be selected, but if it has to be _something_, then the nearest box is fine.) If they wanted to highlight the last sentence beginning with "We observe...", they would likely do a long-press-and-pan gesture similar to this one:

![image](https://github.com/user-attachments/assets/20e18ffb-cbab-44a8-a648-8478282a8ea6)

So a gesture like this could be made to highlight the text ending in "...network."? That would be a vast UX improvement I think, because ending the gesture on the word itself is rather inconvenient using a finger, as the finger or hand can easily obstruct the view of the text one is trying to highlight.

**Comment by hius07** [comment]
You can try it (draft, not polished).
[koptinterface.lua.txt](https://github.com/user-attachments/files/17965850/koptinterface.lua.txt)

For one-column pdf it should work well, but for two- or more columns formatting it is not good.
And we cannot find out how the text is formatted, we have just boxes with their coordinates.

**Comment by jonnyl2** [comment]
I can confirm it works very well on many pdfs; the paragraph endings select as expected, when panning along or to the edge of a line.

However, the attached book seemed to throw it off completely, even on pages without embedded images or unusual formatting. The highlights often started at the other end of the page when panning. (Selections are working fine with the release version.)

Perhaps it can be helpful with ironing out the problems:
[alice-in-wonderland.pdf](https://github.com/user-attachments/files/17969557/alice-in-wonderland.pdf)

**Comment by hius07** [comment]
[koptinterface.lua.txt](https://github.com/user-attachments/files/17974457/koptinterface.lua.txt)
Works good with this book but
-slow (because it sorts page text boxes according to their 'y' coordinate)
-does not work with other books with 2 columns pages.
So, I still do not see the universal solution.

**Comment by hius07** [comment]
We can speed up by some caching and introduce a setting to choose a highlight method. But it would be a manual settings.
Not sure it's a good way.

**Comment by jonnyl2** [comment]
Ok, considering the trade-offs required it may be best to just keep it the way it is. It works quite well for all kinds of layouts.
